sql/schemachanger: Add support for storing roles in policies

spilchen opened the pull request:
A previous commit introduced basic support for CREATE/DROP POLICY. This commit extends that functionality by storing roles in the policy descriptor as strings.

The roles are modelled as separate DSC elements, making it easier to modify the roles in a policy, which will be supported with ALTER POLICY.

Epic: CRDB-11724
Informs: #136696
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139020)
<!-- Reviewable:end -->

Dedej-Bergin reviewed (approved):
Nice work! Just one minor thought/comment.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-:-OG_iqyE0iT687rejkKi:bodrsvk)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OG_j6d_6FWWUJ_Bu1Zk:-OG_j6d_6FWWUJ_Bu1Zl:b-qsqax7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f4e6f614571e87ef34ebafab33ad4402b51d6ef/pkg/sql/catalog/tabledesc/validate.go#L2101)):*
> ```Go
> 		rolesInUse := make(map[string]struct{}, len(p.RoleNames))
> 		for i, roleName := range p.RoleNames {
> 			if _, found := rolesInUse[roleName]; found {
> ```

This validatePolicies() method is getting kind of big, also this `if` nested inside a `for` that's nested inside another `for` loop is making me think that maybe we should break this up.

Maybe we could split some of this up into two additional methods `validatePolicyRoles()` and we can have `validatePolicyNameAndID()`

`validatePolicyRoles()`

```
if len(p.RoleNames) == 0 {
	return errors.AssertionFailedf(
		"policy %q has no roles defined", p.Name)
}
rolesInUse := make(map[string]struct{}, len(p.RoleNames))
		for i, roleName := range p.RoleNames {
			if _, found := rolesInUse[roleName]; found {
				return errors.AssertionFailedf(
					"policy %q contains duplicate role name %q", p.Name, roleName)
			}
			rolesInUse[roleName] = struct{}{}
			// The public role, if included, must always be the first entry in the
			// role names slice.
			if roleName == username.PublicRole && i > 0 {
				return errors.AssertionFailedf(
					"the public role must be the first role defined in policy %q", p.Name)
			}
		}
```

`validatePolicyNameAndID()`

```
		if p.ID == 0 {
			return errors.AssertionFailedf(
				"policy ID was missing for policy %q",
				p.Name)
		} else if p.ID >= desc.NextPolicyID {
			return errors.AssertionFailedf(
				"policy %q has ID %d, which is not less than the NextPolicyID value %d for the table",
				p.Name, p.ID, desc.NextPolicyID)
		}
		if p.Name == "" {
			return pgerror.Newf(pgcode.Syntax, "empty policy name")
		}
		if otherID, found := names[p.Name]; found && p.ID != otherID {
			return pgerror.Newf(pgcode.DuplicateObject,
				"duplicate policy name: %q", p.Name)
		}
		names[p.Name] = p.ID
		if other, found := idToName[p.ID]; found {
			return pgerror.Newf(pgcode.DuplicateObject,
				"policy ID %d in policy %q already in use by %q",
				p.ID, p.Name, other)
		}
		idToName[p.ID] = p.Name
```

Let me know what you think.

 


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139020)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin)

___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OG_j6d_6FWWUJ_Bu1Zk:-OGaBNXC0ZfvBiWPsc_f:b2jdy18) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f4e6f614571e87ef34ebafab33ad4402b51d6ef/pkg/sql/catalog/tabledesc/validate.go#L2101)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

This validatePolicies() method is getting kind of big, also this `if` nested inside a `for` that's nested inside another `for` loop is making me think that maybe we should break this up.

Maybe we could split some of this up into two additional methods `validatePolicyRoles()` and we can have `validatePolicyNameAndID()`

`validatePolicyRoles()`

```
if len(p.RoleNames) == 0 {
	return errors.AssertionFailedf(
		"policy %q has no roles defined", p.Name)
}
rolesInUse := make(map[string]struct{}, len(p.RoleNames))
		for i, roleName := range p.RoleNames {
			if _, found := rolesInUse[roleName]; found {
				return errors.AssertionFailedf(
					"policy %q contains duplicate role name %q", p.Name, roleName)
			}
			rolesInUse[roleName] = struct{}{}
			// The public role, if included, must always be the first entry in the
			// role names slice.
			if roleName == username.PublicRole && i > 0 {
				return errors.AssertionFailedf(
					"the public role must be the first role defined in policy %q", p.Name)
			}
		}
```

`validatePolicyNameAndID()`

```
		if p.ID == 0 {
			return errors.AssertionFailedf(
				"policy ID was missing for policy %q",
				p.Name)
		} else if p.ID >= desc.NextPolicyID {
			return errors.AssertionFailedf(
				"policy %q has ID %d, which is not less than the NextPolicyID value %d for the table",
				p.Name, p.ID, desc.NextPolicyID)
		}
		if p.Name == "" {
			return pgerror.Newf(pgcode.Syntax, "empty policy name")
		}
		if otherID, found := names[p.Name]; found && p.ID != otherID {
			return pgerror.Newf(pgcode.DuplicateObject,
				"duplicate policy name: %q", p.Name)
		}
		names[p.Name] = p.ID
		if other, found := idToName[p.ID]; found {
			return pgerror.Newf(pgcode.DuplicateObject,
				"policy ID %d in policy %q already in use by %q",
				p.ID, p.Name, other)
		}
		idToName[p.ID] = p.Name
```

Let me know what you think.
</blockquote></details>

Good idea. It's probably a good idea to split up that function. I pushed a change to make the functions smaller. Let me know what you think.


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139020)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OG_j6d_6FWWUJ_Bu1Zk:-OGaN73L3opZRvPBfbj6:bsqv41m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f4e6f614571e87ef34ebafab33ad4402b51d6ef/pkg/sql/catalog/tabledesc/validate.go#L2101)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Good idea. It's probably a good idea to split up that function. I pushed a change to make the functions smaller. Let me know what you think.
</blockquote></details>

Very nice, I like it!


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
this lgtm! i just had a few comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-:-OGcca21BGnqCJLBnLXf:biz5x2p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_policy.go` line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OGcbtQ71X-Zc-lOdREQ:-OGcbtQ71X-Zc-lOdRER:b263lcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90749b01a1ca627c17d3e12c4bf56b7686a31af/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_policy.go#L99)):*
> ```Go
> 	}
> 	for _, role := range n.Roles {
> 		switch role.RoleSpecType {
> ```

i believe this logic is correct, but i think the best approach here is to re-use the `decodeusername.FromRoleSpec` function, like we do in other places: https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_schema.go#L77-L79
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OGcatJLAWmprTPxFr9Z:-OGcatJLAWmprTPxFr9_:b-cggpu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90749b01a1ca627c17d3e12c4bf56b7686a31af/pkg/sql/catalog/tabledesc/validate.go#L2110)):*
> ```Go
> 
> // validatePolicyRoles will validate the roles that are in one policy.
> func (desc *wrapper) validatePolicyRoles(p *descpb.PolicyDescriptor) error {
> ```

In addition to the validation here, we also should update the `ValidateRolesInDescriptor` function: https://github.com/cockroachdb/cockroach/blob/0cdbb75acbc4ca136dcc26ed1fd130b3c7d4eadf/pkg/sql/catalog/validate.go#L198

That one is able to verify that the roles in the descriptor actually exist.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139020)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin and @rafiss)

___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OGcatJLAWmprTPxFr9Z:-OGevEWI1KcQb1m5L_PY:bo12qzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90749b01a1ca627c17d3e12c4bf56b7686a31af/pkg/sql/catalog/tabledesc/validate.go#L2110)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

In addition to the validation here, we also should update the `ValidateRolesInDescriptor` function: https://github.com/cockroachdb/cockroach/blob/0cdbb75acbc4ca136dcc26ed1fd130b3c7d4eadf/pkg/sql/catalog/validate.go#L198

That one is able to verify that the roles in the descriptor actually exist.
</blockquote></details>

Good idea! I couldn’t directly call `ValidateRolesInDescriptor` for the `PolicyDescriptor` since it doesn’t implement the descriptor interface. Instead, I extracted relevant parts and created a standalone function, `ValidateRole`, which is now used to validate each role in the policies.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_policy.go` line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OGcbtQ71X-Zc-lOdREQ:-OGe4fNo2BJRBBGo9_2O:btuqo2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90749b01a1ca627c17d3e12c4bf56b7686a31af/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_policy.go#L99)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i believe this logic is correct, but i think the best approach here is to re-use the `decodeusername.FromRoleSpec` function, like we do in other places: https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_schema.go#L77-L79
</blockquote></details>

I wasn't aware of this function. It cleans this code up.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139020)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin and @spilchen)

___
*[`pkg/sql/catalog/validate.go` line 239 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OGjfmUE7fMJoVj57w3w:-OGjfmUE7fMJoVj57w3x:b560go1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5fbe228857aa16504279d3e27067077d62cf496/pkg/sql/catalog/validate.go#L239)):*
> ```Go
> 		}); err != nil {
> 			return err
> 		}
> ```

i didn't fully understand why we needed to introduce the new ValidateRole function. can't the loop over policies be in this function, like this:
```
	if tbDesc, ok := descriptor.(TableDescriptor); ok {
		for _, p := range tbDesc.GetPolicies() {
			for _, r := range p.RoleNames {
				exists, err := RoleExists(r)
				if err != nil {
					return err
				}
				if !exists {
					return errors.AssertionFailedf("descriptor %q (%d) has a policy (%s) for a role %q that doesn't exist",
						descriptor.GetName(),
						descriptor.GetID(),
						p.Name,
						r,
					)
				}
			}
		}
	}
```


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Add support for storing roles in policies

A previous commit introduced basic support for CREATE/DROP POLICY. This
commit extends that functionality by storing roles in the policy
descriptor as strings.

The roles are modelled as separate DSC elements, making it easier to
modify the roles in a policy, which will be supported with ALTER POLICY.

Epic: CRDB-11724
Informs: #136696
Release note: None

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139020)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin and @rafiss)

___
*[`pkg/sql/catalog/validate.go` line 239 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-OGjfmUE7fMJoVj57w3w:-OGjpSiC0er45wIgn4-X:bv08pez) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5fbe228857aa16504279d3e27067077d62cf496/pkg/sql/catalog/validate.go#L239)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i didn't fully understand why we needed to introduce the new ValidateRole function. can't the loop over policies be in this function, like this:
```
	if tbDesc, ok := descriptor.(TableDescriptor); ok {
		for _, p := range tbDesc.GetPolicies() {
			for _, r := range p.RoleNames {
				exists, err := RoleExists(r)
				if err != nil {
					return err
				}
				if !exists {
					return errors.AssertionFailedf("descriptor %q (%d) has a policy (%s) for a role %q that doesn't exist",
						descriptor.GetName(),
						descriptor.GetID(),
						p.Name,
						r,
					)
				}
			}
		}
	}
```
</blockquote></details>

Wow, this is way easier than what I did. Not sure how I missed that. I corrected this in the last commit.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> Reviewed 28 of 30 files at r1, 1 of 1 files at r2, 1 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139020#-:-OGk8lohDZgrNa41uXyA:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
